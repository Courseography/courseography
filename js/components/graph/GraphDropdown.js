import React from "react"
import PropTypes from "prop-types"

export default class GraphDropdown extends React.Component {
  render() {
    let className = "hidden"
    let graphTabLeft = 0
    if (this.props.graphs.length !== 0 && document.querySelector("#nav-graph")) {
      let navGraph = document.querySelector("#nav-graph")
      if (this.props.graphs.length === 0) {
        navGraph.classList.remove("show-dropdown-arrow")
      } else {
        if (!navGraph.classList.contains("show-dropdown-arrow")) {
          navGraph.classList.add("show-dropdown-arrow")
        }
        if (this.props.showGraphDropdown) {
          graphTabLeft = navGraph.getBoundingClientRect().left
          className = "graph-dropdown-display"
        }
      }
    }

    return (
      <ul
        className={className}
        onMouseMove={this.props.onMouseMove}
        onMouseLeave={this.props.onMouseLeave}
        data-testid={"test-graph-dropdown"}
        style={{ left: graphTabLeft }}
      >
        {this.props.graphs.map((graph, i) => {
          return (
            <li
              key={i}
              className="graph-dropdown-item"
              onClick={() => this.props.updateGraph(graph.title)}
              data-testid={"test-graph-" + i}
            >
              {graph.title}
            </li>
          )
        })}
      </ul>
    )
  }
}

GraphDropdown.defaultProps = {
  graphs: [],
}

GraphDropdown.propTypes = {
  showGraphDropdown: PropTypes.bool,
  onMouseMove: PropTypes.func,
  onMouseLeave: PropTypes.func,
  graphs: PropTypes.array,
  updateGraph: PropTypes.func,
}
