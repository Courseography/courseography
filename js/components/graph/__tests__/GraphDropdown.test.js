import React from "react"
import TestContainer from "./TestContainer"
import GraphDropdown from "../GraphDropdown"
import { shallow } from "enzyme"
import { fireEvent } from "@testing-library/react"

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms))
}

describe("GraphDropdown", () => {
  it("should match shallow snapshot", () => {
    const graphDropdownProps = {
      onMouseMove: jest.fn(),
      onMouseLeave: jest.fn(),
      showGraphDropdown: false,
      graphs: [],
      updateGraph: jest.fn(),
    }
    const component = shallow(<GraphDropdown {...graphDropdownProps} />)
    expect(component).toMatchSnapshot()
  })
  it("should appear when hovering over the graph tab and be hidden before", async () => {
    const mouseEnter = new MouseEvent("mouseenter", {
      bubbles: false,
      cancelable: false,
    })

    const container = await TestContainer.build()
    const graphNav = container.getByText("Graph")
    const graphDropdown = container.getByTestId("test-graph-dropdown")
    expect(graphDropdown.classList.contains("hidden")).toBe(true)
    fireEvent(graphNav, mouseEnter)
    expect(graphDropdown.classList.contains("graph-dropdown-display")).toBe(true)
  })

  it("should disappear a second after the cursor isn't hovered on the dropdown", async () => {
    const mouseEnter = new MouseEvent("mouseenter", {
      bubbles: false,
      cancelable: false,
    })

    const container = await TestContainer.build()
    const graphNav = container.getByText("Graph")
    const graphDropdown = container.getByTestId("test-graph-dropdown")
    expect(graphDropdown.classList.contains("hidden")).toBe(true)
    fireEvent(graphNav, mouseEnter)
    expect(graphDropdown.classList.contains("graph-dropdown-display")).toBe(true)

    fireEvent.mouseOut(graphDropdown)
    await sleep(600)
    expect(graphDropdown.classList.contains("hidden")).toBe(true)
  })

  it("should disappear a second after the cursor isn't hovered on the graph tab", async () => {
    const mouseEnter = new MouseEvent("mouseenter", {
      bubbles: false,
      cancelable: false,
    })
    const mouseLeave = new MouseEvent("mouseleave", {
      bubbles: false,
      cancelable: false,
    })

    const container = await TestContainer.build()
    const graphNav = container.getByText("Graph")
    const graphDropdown = container.getByTestId("test-graph-dropdown")
    expect(graphDropdown.classList.contains("hidden")).toBe(true)
    fireEvent(graphNav, mouseEnter)
    expect(graphDropdown.classList.contains("graph-dropdown-display")).toBe(true)

    fireEvent(graphNav, mouseLeave)
    await sleep(600)
    expect(graphDropdown.classList.contains("hidden")).toBe(true)
  })
})
