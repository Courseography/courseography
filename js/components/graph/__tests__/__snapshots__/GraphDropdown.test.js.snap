// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphDropdown should match shallow snapshot 1`] = `
<ul
  className="hidden"
  data-testid="test-graph-dropdown"
  onMouseLeave={[MockFunction]}
  onMouseMove={[MockFunction]}
  style={
    Object {
      "left": 0,
    }
  }
/>
`;
